### PR TITLE
2021-01-26 - When should we automate test discussion summary

### PR DESCRIPTION
**Pull request by adichourasiya**
Hi team,
    
    publishing this article as discussed, to summarize our discussion on when we should automate test and when should we avoid it. It's a consolidated version, I'll put the video in next commit for elaborated /conversational version.

### CODE REVIEW

**Comment by wpschaub** [comment]
@adichourasiya you can only specify one author in the meta data, which links back to the authors page. At the start of the article I would add a Authored by: X, Y, and Z.

Please use kebab notation for filename and slug, i.e. when-should-we-automate-tests, not When should we automate tests.

@wsbctechnicalblog we may have to update our 101 blog post to add this type of guidance.

**Review by wsbctechnicalblog — CHANGES_REQUESTED** [review]
My alarm bells went off when I read that tests in support of hotfixes should not be automated!!!

**Comment by adichourasiya** [comment]
> My alarm bells went off when I read that tests in support of hotfixes should not be automated!!!

Good catch, I changed it to say "Urgent non-regression hotfix can/may be manually verified."  Regression test for hot fix scenarios should still be automated.

**Comment by adichourasiya** [comment]
> @adichourasiya you can only specify one author in the meta data, which links back to the authors page. At the start of the article I would add a Authored by: X, Y, and Z.
> 
> Please use kebab notation for filename and slug, i.e. when-should-we-automate-tests, not When should we automate tests.
> 
> @wsbctechnicalblog we may have to update our 101 blog post to add this type of guidance.

Updated in latest commit. This is good information , I'll update in 101.

File: content/when-should-we-automate-tests.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+Slug: when-should-we-automate-tests
+Author: Aditya Chourasiya
+
+Summary: This article is gist of brainstorming session among QA team(Mohita Kataria, Midhi Naidu and Aditya Chourasiya) at WorkSafeBC
+
+   Automation is advised when tests are **prone to human error** or **complex** or if they need **repeating**, **reporting** , **execution speed** and ofcourse fetch good **return on investment**. 
+   
+   On the conterary, test cases for aesthetics, accessibility, one-time, urgent non-regression(hotfixes), exploratory, with unpredictable results and/or with less return on investment,  should not be automated. 
+   
+<img src="../images/chapline-automation.gif">
```
Comment (wpschaub) [inline content/when-should-we-automate-tests.md:14]: @adichourasiya  can you change this line to:  Chaplain

File: content/when-should-we-automate-tests.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+ These are few reasons not to automate (not in order)
+- **Aesthetics** - User experience , look and feel, colour and images verification require complex human context and interpretation.
+- **Accessibility** - Similar to aesthetics, accessibility, usablity, language, grammer etc should be verified manually
+- **One-time** - Urgent non-regression hotfix, non-critical, or very small changes can/may be manually verified  
+- **Unpredictable results** - Exploratory, time delay tests, random tests can/will rely on manual execution
+- **Inadequate return on investment** - Automated test for short lived products, small scale development, duplicate tests should be avoided 
+
+# Summary and invite 
+While ability and enthusiasm to automate is important, it is more important to know when to automate and avoid [gold plating](https://en.wikipedia.org/wiki/Gold_plating_(project_management)). I will link the video for discussion in next commit. Thank you for reading. Please feel free to submit a pull request to suggest any changes.
```
Comment (wpschaub) [inline content/when-should-we-automate-tests.md]: @adichourasiya please add two empty lines at the end - we have experienced formatting issues if they are not present.

File: content/when-should-we-automate-tests.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- **Critical path** - If the user action sequence / path or application function is critical to your business, it's recommended to automate.
+- **Prone to human error** - large number of data fields, complex technical steps like Database test, Security tests should be automated.
+- **Reporting** - Legal, audit, low confidence or high stake tests should be automated for reporting and logs.
+- **Execution speed** - parallel and automated execution can help run all sort of test much faster, however planning, automation, triage may take more time depending on type of tests.
+- **Return on investment** - While there is no exact number, as a rule of thumb anytime you're saving more money with automation, you should.
+
+ These are few reasons not to automate (not in order)
+- **Aesthetics** - User experience , look and feel, colour and images verification require complex human context and interpretation.
+- **Accessibility** - Similar to aesthetics, accessibility, usablity, language, grammer etc should be verified manually
+- **One-time** - Urgent non-regression hotfix, non-critical, or very small changes can/may be manually verified
```
Comment (wpschaub) [inline content/when-should-we-automate-tests.md:32]: @adichourasiya Are we sure with this? Would a hotfix not be treated like a feature, in which case we should have automated unit and regression testing?

File: content/when-should-we-automate-tests.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- **Prone to human error** - large number of data fields, complex technical steps like Database test, Security tests should be automated.
+- **Reporting** - Legal, audit, low confidence or high stake tests should be automated for reporting and logs.
+- **Execution speed** - parallel and automated execution can help run all sort of test much faster, however planning, automation, triage may take more time depending on type of tests.
+- **Return on investment** - While there is no exact number, as a rule of thumb anytime you're saving more money with automation, you should.
+
+ These are few reasons not to automate (not in order)
+- **Aesthetics** - User experience , look and feel, colour and images verification require complex human context and interpretation.
+- **Accessibility** - Similar to aesthetics, accessibility, usablity, language, grammer etc should be verified manually
+- **One-time** - Urgent non-regression hotfix, non-critical, or very small changes can/may be manually verified  
+- **Unpredictable results** - Exploratory, time delay tests, random tests can/will rely on manual execution
```
Comment (wpschaub) [inline content/when-should-we-automate-tests.md:33]: @adichourasiya are these "flaky" tests?

File: content/when-should-we-automate-tests.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+Title: When should we automate tests
+Date: 2022-01-01
```
Comment (wpschaub) [inline content/when-should-we-automate-tests.md]: @wsbctechnicalblog you will have to update this date B4 publishing.

**Review by wpschaub — CHANGES_REQUESTED** [review]
A few minor tweaks requested.

File: content/when-should-we-automate-tests.md (hunk 3 above)
Comment (adichourasiya) [reply content/when-should-we-automate-tests.md:32]: I believe we're saying the same thing, hotfix will have automated unit, integration and UI test part of regression suite, however if these tests are falling outside category of regression, (something quick and dirty, one time verification) , they need not to be automated.

**Review by wpschaub — CHANGES_REQUESTED** [review]
@adichourasiya similar to your other PR, please move your images from /content/images folder to /images folder so that they light up. Other than that we are good to launch this one.

File: content/when-should-we-automate-tests.md (hunk 5 above)
Comment (wpschaub) [reply content/when-should-we-automate-tests.md]: @wsbctechnicalblog I have fixed the image issue for @adichourasiya and added two empty lines at the end. The post is good to go and will be our first QA post - I assume that @adichourasiya will add the promised video link in future.

**Comment by wpschaub** [comment]
@wsbctechnicalblog I have fixed the image issue for @adichourasiya and added two empty lines at the end. The post is good to go and will be our first QA post - I assume that @adichourasiya will add the promised video link in future.

**Review by wpschaub — DISMISSED** [review]

**Review by wpschaub — DISMISSED** [review]
After a wave of tweaks, we are good to go.

**Comment by wpschaub** [comment]
@wsbctechnicalblog @adichourasiya we can publish this ASAP as part of Aaron's handover!

**Pull request merged**